src/Symfony/Component/PropertyInfo/WriteMutator.php

```diff
@@ -0,0 +1,80 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\PropertyInfo;
+
+/**
+ * Writes mutator tell how to write to a property.
+ *
+ * @author Joel Wurtz <jwurtz@jolicode.com>
+ */
+final class WriteMutator
+{
+    public const TYPE_METHOD = 1;
+    public const TYPE_PROPERTY = 2;
+    public const TYPE_ARRAY_DIMENSION = 3;
+    public const TYPE_ADDER_REMOVER_METHOD = 4;
+    public const TYPE_CONSTRUCTOR = 5;
+
+    private $type;
+    private $name;
+    private $adderName;
+    private $removerName;
+    private $private;
+    private $constructorParameter;
+    private $static;
+
+    public function __construct(int $type, ?string $name, ?string $adderName, ?string $removerName, bool $private = false, bool $static = false, \ReflectionParameter $constructorParameter = null)
+    {
+        $this->type = $type;
+        $this->name = $name;
+        $this->adderName = $adderName;
+        $this->removerName = $removerName;
+        $this->private = $private;
+        $this->constructorParameter = $constructorParameter;
+        $this->static = $static;
+    }
+
+    public function getType(): int
+    {
+        return $this->type;
+    }
+
+    public function getName(): ?string
+    {
+        return $this->name;
+    }
+
+    public function getAdderName(): ?string
+    {
+        return $this->adderName;
+    }
+
+    public function getRemoverName(): ?string
+    {
+        return $this->removerName;
+    }
+
+    public function isPrivate(): bool
+    {
+        return $this->private;
+    }
+
+    public function isStatic(): bool
+    {
+        return $this->static;
+    }
+
+    public function getConstructorParameter(): ?\ReflectionParameter
+    {
+        return $this->constructorParameter;
```

Comment:
this seems a bit fragile to me. should we check the type and throw an exception if its not TYPE_CONSTRUCTOR? then we could make the return type of this non-nullable. (and also validate in the constructor)

Reply:
Why not, would prefer to not have a lot of logic here, for me this class is mainly a helper for IDE and stuff that avoid using an array like in the PropertyAccess component.

Adding exception can have some bias in other component where we would need to take care of this (more maintenance)

Reply:
if you want to be sure that your code won't fatal in random places, having optional return value is a huge pain as you need to check if you actually got something. (thats why phpstan will complain)

imho an exception when you do something you are not supposed to do would be more helpful.

i agree we should not have too much logic on these, but we have models that cover different cases and we should make sure that the cases do not get mixed up. the super-duper OO alternative would be some marker interface and an implementation for each type, but that seems overcomplicated to me, and makes it harder to see which types exist.